File: core/src/main/java/org/lsposed/lspd/hooker/CrashDumpHooker.java
package org.lsposed.lspd.hooker;

import org.lsposed.lspd.impl.LSPosedBridge;
import org.lsposed.lspd.util.Utils.Log;

import io.github.libxposed.api.XposedInterface;

public class CrashDumpHooker implements XposedInterface.Hooker {

    public static void before(XposedInterface.BeforeHookCallback callback) {
        try {
            var e = (Throwable) callback.getArgs()[0];
            LSPosedBridge.log("Crash unexpectedly: " + Log.getStackTraceString(e));
        } catch (Throwable ignored) {
        }
    }
}

File: xposed/src/main/kotlin/org/matrix/vector/impl/hookers/CrashDumpHooker.kt
package org.matrix.vector.impl.hookers

import android.util.Log
import io.github.libxposed.api.XposedInterface
import org.matrix.vector.impl.VectorLogBridge

/**
 * Intercepts uncaught exceptions in the framework to provide diagnostic logging before the process
 * completely terminates.
 */
object CrashDumpHooker : XposedInterface.Hooker {
    override fun intercept(chain: XposedInterface.Chain): Any? {
        try {
            val throwable = chain.args.firstOrNull() as? Throwable
            if (throwable != null) {
                VectorLogBridge.log(
                    Log.ERROR,
                    "NPatchCrash",
                    "Uncaught exception in ${Thread.currentThread().name}",
                    throwable,
                )
            }
        } catch (ignored: Throwable) {}
        return chain.proceed()
    }
}
